File: learntoprogram/InnerClass.java
/**
* Description: 内部类、静态内部类
*/

class Outer
{
    int x = 4;
    class Inner
    {
        int y = 5;
        void show()
        {
            int z = 6;
            System.out.println("show...innershowvalue" + z);//默认优先采用局部变量
            System.out.println("show...innervalue" + this.y);//其次向上搜索类成员变量
            System.out.println("show...outervalue" + Outer.this.x);//再向上搜索外部类的成员变量，内部类持有外部类的引用,Outer.this可省略
        }
    }

    static class InnerStatic //静态内部类
    {
        void show()
        {
            System.out.println("show...static");
        }
    }
}

class InnerClass
{
    public static void main(String [] args)
    {
        //普通内部类初始化需要两步
        Outer o = new Outer();
        Outer.Inner i = o.new Inner();
        i.show();    

        //静态内部类可以单独初始化,但不能访问非静态资源
        Outer.InnerStatic si = new Outer.InnerStatic();
        si.show();
    }
}

File: learntoprogram/InnerClass.py
#!/usr/bin/python
#coding:utf-8
##################################
"""
Description: 内部类、局部内部类
Version: v1.0
"""
##################################

class Outer(object):
    def __init__(self):
        self.x = 4

    class Inner(object):
        def __init__(self):
            self.y = 5
            self.out = Outer() #调用外部类的实例变量或实例函数时，需先将外部类的实例传入内部类
            
        def show(self):
            z = 6
            print "show...innershowV" + str(z)#调用局部变量
            print "show...innerV" + str(self.y)#调用内部类实例变量
            print "show...outerV" + str(self.out.x)#调用外部类实例变量

    #返回一个局部内部类的实例
    def method(self):
        x = 10
        #继承内部类的局部内部类
        class InnerLocal(Outer.Inner): 
            print x

        il = InnerLocal()
        return il

if __name__ == '__main__':
    o = Outer()
    i = o.Inner()
    i.show()
    il = o.method()
    il.show()
